pkg/cache/v3/delta.go

```diff
@@ -0,0 +1,103 @@
+//   limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/envoyproxy/go-control-plane/pkg/log"
+	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
+)
+
+// Respond to a delta watch with the provided snapshot value
+func respondDelta(request *DeltaRequest, value chan DeltaResponse, st *stream.StreamState, snapshot Snapshot, log log.Logger) *RawDeltaResponse {
+	resp, err := createDeltaResponse(request, st, snapshot)
+	if err != nil {
+		if log != nil {
+			log.Errorf("Error creating delta response: %v", err)
+		}
+		return nil
+	}
+
+	// Only send a response if there were changes
+	if len(resp.Resources) > 0 || len(resp.RemovedResources) > 0 {
+		if log != nil {
+			log.Debugf("node: %s, sending delta response with resources: %v removed resources %v wildcard: %t",
+				request.GetNode().GetId(), resp.Resources, resp.RemovedResources, st.IsWildcard)
+		}
+		value <- resp
+		return resp
+	}
+	return nil
+}
+
+// snapshot.GetResources(watch.Request.TypeUrl),
+// snapshot.GetVersion(watch.Request.TypeUrl),
+
+func createDeltaResponse(request *DeltaRequest, st *stream.StreamState, snapshot Snapshot) (*RawDeltaResponse, error) {
+	resources := snapshot.GetResources((request.TypeUrl))
+
+	// variables to build our response with
+	nextVersionMap := make(map[string]string)
+	filtered := make([]types.Resource, 0, len(resources))
+	toRemove := make([]string, 0)
+
+	// Wildcard can happen through CDS/LDS. If this is done we want to track all resources as well as track them in the version map
+	if st.IsWildcard {
+		for name, r := range resources {
+			// Since we've already precomputed the version hashes of the new snapshot,
+			// we can just set it here to be used for comparison later
+			version := snapshot.GetVersionMap()[request.TypeUrl][name]
+			fmt.Println("new version " + version)
+			nextVersionMap[name] = version
+			prevVersion, found := st.ResourceVersions[name]
+			if !found || (prevVersion != nextVersionMap[name]) {
+				filtered = append(filtered, r)
+			}
+		}
+	} else {
+		// Reply only with the requested resources. Envoy may ask each resource
+		// individually in a separate stream. It is ok to reply with the same version
+		// on separate streams since requests do not share their response states.
+		for name, prevVersion := range st.ResourceVersions {
+			if r, ok := resources[name]; ok {
+				nextVersion := snapshot.GetVersionMap()[request.TypeUrl][name]
+				if prevVersion != nextVersion {
+					filtered = append(filtered, r)
+				}
+				nextVersionMap[name] = nextVersion
+			} else {
+				// if prevVersion == "" this means that the resourse was already removed or doesn't yet exist on the client.
+				if prevVersion != "" {
+					toRemove = append(toRemove, name)
+				}
+
+				// the resource is gone but we keep tracking it in the
+				// version map so we can detect an update if the resource comes back
```

Comment:
Why does this matter? Why not just nil it out?

Reply:
Well the map is of type `map[string]string` so we can't nil it out or go will throw a type error:
```go
nextVersionMap[name] = nil
```
yields
```
cannot use nil (untyped nil value) as string value in assignmentcompilerIncompatibleAssign
```

But we could do something like:
```go
delete(nextVersionMap, name)
```
and that would nil out that entry. I guess we can just remove that because if the client does ever want that resource again it'll just ask for it correct?